### PR TITLE
Update and test `EnumerableMap`

### PR DESCRIPTION
**Pull request by NouDaimon**
Contains:

- tests for EnumerableMap
- fix for `_at(AddressToAddressMap storage map, uint256 index)` function

Initially only for `AddressToAddressMap`, once passed initial scrutiny, the `UintToAddressMap` tests will be added.

### CODE REVIEW

File: contracts/data/EnumerableMap.sol
Comment (NouDaimon) [inline contracts/data/EnumerableMap.sol:43]: This is not an elegant solution. 

I'll have a look at what else can be done.

Comment (ItsNickBarry) [reply contracts/data/EnumerableMap.sol:43]: I don't know why there was an assembly block here.  All we need is a simple cast: `return (address(uint160(uint256(key))), address(uint160(uint256(value))));`

File: test/data/EnumerableMap.ts
Comment (ItsNickBarry) [inline test/data/EnumerableMap.ts:37]: In several places you have multiple similar calls.  I don't think it adds much value.  One is enough.

File: test/data/EnumerableMap.ts
Comment (ItsNickBarry) [inline test/data/EnumerableMap.ts:42]: I don't like the use of `deep.equal` here.  Better to destructure the return data and compare each piece.
```js
const [key, value] = await instance.callStatic['at(uint256)'](0);
expect(key).to.equal(...);
```

Comment (ItsNickBarry) [reply test/data/EnumerableMap.ts:42]: Also, use `callStatic` when calling view and pure functions.

File: test/data/EnumerableMap.ts
Comment (ItsNickBarry) [inline test/data/EnumerableMap.ts:101]: This is a good case for making multiple calls.  I would probably put them in a loop, though doing that is dangerous because if a loop has 0 iterations, the issue won't show up as a test failure.

File: test/data/EnumerableMap.ts
Comment (0xCourtney) [inline test/data/EnumerableMap.ts:91]: perhaps you can add a case or logic that shows adding an entry of an existing key does not increase the length

```js
...
await instance['set(address,address)'](addressThree, addressSix);
expect(await instance['length()']()).to.equal(3);

await instance['set(address,address)'](addressThree, addressSeven);
expect(await instance['length()']()).to.equal(3);
...
```

**Review by 0xCourtney — COMMENTED** [review]
Good job getting this going! Does it make sense to add [`toArray`](https://github.com/solidstate-network/solidstate-solidity/blob/EnumerableMapTests/contracts/data/EnumerableSet.sol#L153-L168) as we did for `EnumerableSet`?

**Comment by ItsNickBarry** [comment]
@0xCourtney It's not immediately clear what `toArray` would return.  Is it possible to return an array of tuples?  If not, maybe two separate functions for keys and values?

**Comment by 0xCourtney** [comment]
> @0xCourtney It's not immediately clear what `toArray` would return. Is it possible to return an array of tuples or array of arrays? If not, maybe two separate functions for keys and values?

Great question, I think supporting `toArray` which returns an array of tuples makes sense as it keeps the interface consistent with the other data structures.

```solidity
struct AddressToAddress {
    address key;
    address value;
}

function toArray(AddressToAddressMap storage map)
    internal
    view
    returns (AddressToAddress[] memory)
{
    uint256 len = _length(map._inner);
    AddressToAddress[] memory arr = new AddressToAddress[](len);

    unchecked {
        for (uint256 index; index < len; ++index) {
            (address key, address value) = at(map, index);
            arr[index] = AddressToAddress({ key: key, value: value });
        }
    }

    return arr;
}

// [
//   [
//     '0x0000000000000000000000000000000000000064',
//     '0x000000000000000000000000000000000000012c',
//     key: '0x0000000000000000000000000000000000000064',
//     value: '0x000000000000000000000000000000000000012c'
//   ],
//   [
//     '0x00000000000000000000000000000000000000C8',
//     '0x000000000000000000000000000000000000012c',
//     key: '0x00000000000000000000000000000000000000C8',
//     value: '0x000000000000000000000000000000000000012c'
//   ],
//   [
//     '0x000000000000000000000000000000000000012c',
//     '0x00000000000000000000000000000000000000C8',
//     key: '0x000000000000000000000000000000000000012c',
//     value: '0x00000000000000000000000000000000000000C8'
//   ]
// ]
```


Adding `keys` and `values` (not particular on the naming) which return an array of keys and an array of values, respectively, would probably be more useful for the dev.

File: test/data/EnumerableMap.ts
Comment (0xCourtney) [inline test/data/EnumerableMap.ts:130]: We should include a test which demonstrates that calling `set` for a key that already exists overwrites the value.

```js
it('overwrites the address value at address key', async () => {
  await instance['set(address,address)'](addressOne, addressTwo);
  await instance['set(address,address)'](addressOne, addressThree);

  expect(await instance['get(address)'](addressOne)).to.eq(
    addressThree,
  );
});
```

**Review by 0xCourtney — CHANGES_REQUESTED** [review]

**Comment by ItsNickBarry** [comment]
@0xCourtney Your example returns a struct.  Do you know if Solidity supports tuple arrays?  If it doesn't, I'd favor returning two separate arrays over defining a single-use struct.

**Comment by 0xCourtney** [comment]
> @0xCourtney Your example returns a struct. Do you know if Solidity supports tuple arrays? If it doesn't, I'd favor returning two separate arrays over defining a single-use struct.

No, tuples are not a proper type and are only accessible to Solidity internally according to [this](https://docs.soliditylang.org/en/latest/control-structures.html#destructuring-assignments-and-returning-multiple-values). Likewise, if `toArray` returns a tuple of arrays it would be much cleaner.

**Comment by ItsNickBarry** [comment]
Thanks for the confirmation.  In that case, we'll do it like in my example interface above.

**Review by ItsNickBarry — APPROVED** [review]
Didn't check whether all tests are copy/pasted properly, but coverage is at 100%.

No real reason to use `EnumerableSetTests` as the base branch.  Now we need to wait until that branch is merged before merging this one.

**Review by 0xCourtney — APPROVED** [review]

**Pull request merged**
